django/db/models/query_utils.py

```diff
@@ -95,6 +96,36 @@ def resolve_expression(
         query.promote_joins(joins)
         return clause
 
+    def flatten(self):
+        yield self
+        for child in self.children:
+            yield from child.flatten()
+
+    def check(self, against, using=DEFAULT_DB_ALIAS):
+        """
+        Do a database query to check if the expressions of the Q instance
+        matches against the `against` expressions.
+        """
+        # Avoid circular imports
+        from django.db.models import Value
+        from django.db.models.expressions import RawSQL
+        from django.db.models.sql import Query
+        from django.db.models.sql.constants import SINGLE
+
+        query = Query(None)
+        # `select` needs to be True so that RawSQL expressions have access
+        # to the values
+        select = any((isinstance(expr, RawSQL) for expr in self.children))
```

Comment:
Indeed, I meant `self.flatten` instead of `self.children`. Which made me realize my `Q.flatten()` implementation wasn't actually correct. I pushed a fix for it.